### PR TITLE
ARROW-6605: [C++][Filesystem] Add recursion depth control to fs::Selector

### PR DESCRIPTION
**Pull request by fsaintjacques**

### CODE REVIEW

File: cpp/src/arrow/filesystem/test_util.cc [hunk 1]
```diff
@@ -646,6 +647,114 @@ void GenericFileSystemTest::TestGetTargetStatsSelector(FileSystem* fs) {
...
+
+FileStats Dir(std::string path) {
+  FileStats st;
+  st.set_type(FileType::Directory);
+  st.set_path(path);
+  return st;
+}
+
+// Utility functions to help testing/debugging
+std::ostream& operator<<(std::ostream& os, const FileStats& stats) {
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/test_util.cc]: Why not put this in the public API?

File: cpp/src/arrow/filesystem/test_util.cc [hunk 2]
```diff
@@ -646,6 +647,114 @@ void GenericFileSystemTest::TestGetTargetStatsSelector(FileSystem* fs) {
   ASSERT_RAISES(IOError, fs->GetTargetStats(s, &stats));
 }
 
+using ::testing::ElementsAre;
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/test_util.cc]: Put all utility stuff somewhere near the top of the file?

File: cpp/src/arrow/filesystem/test_util.cc [hunk 3]
```diff
@@ -646,6 +647,114 @@ void GenericFileSystemTest::TestGetTargetStatsSelector(FileSystem* fs) {
...
+  st.set_path(path);
+  return st;
+}
+
+// Utility functions to help testing/debugging
+std::ostream& operator<<(std::ostream& os, const FileStats& stats) {
+  return os << "FileStats(" << stats.type() << ", " << stats.path() << ")";
+}
+
+bool operator==(const FileStats& lhs, const FileStats& rhs) {
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/test_util.cc]: We can probably put this in the public API, but it must compare size and mtime as well.

File: cpp/src/arrow/filesystem/test_util.cc [hunk 4]
```diff
@@ -646,6 +647,114 @@ void GenericFileSystemTest::TestGetTargetStatsSelector(FileSystem* fs) {
...
+                                 Dir("01/02"), Dir("AA"), File("AA/AA.file")));
+
+  s.recursive = true;
+  s.max_recursion = 2;
+  GetSortedStats(fs, s, stats);
+  EXPECT_THAT(stats, ElementsAre(File("00.file"), Dir("01"), File("01/01.file"),
+                                 Dir("01/02"), File("01/02/02.file"), Dir("01/02/03"),
+                                 Dir("AA"), File("AA/AA.file")));
+
+  s.base_dir = "AA";
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/test_util.cc]: Perhaps cut down a bit on the number of similar tests here? For something like S3 the roundtrips are a bit slow.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for adding this.

File: cpp/src/arrow/filesystem/test_util.cc (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/filesystem/test_util.cc]: I think it's better to keep the tests and disable some for s3, if possible

Comment (fsaintjacques) [reply cpp/src/arrow/filesystem/test_util.cc]: I removed 2 redundant checks

**Pull request closed without merging**